Working WIP exhaustive case macro

bobjflong opened the pull request:
The following macro won't even expand in the test!

``` elixir
AdtTest.AdtDefinition.case %AdtTest.AdtDefinition.Foo{}, [
  Foo: fn(x) -> "foo" end
]
```

```
** Not exhaustive!
    expanding macro: AdtTest.AdtDefinition.case/2
    test/adt_test.exs:68: AdtTest."test case"/1
    (elixir) lib/code.ex:363: Code.require_file/2
```

This needs to be cleaned up **a lot**! And the syntax needs some thinking through.

bobjflong committed:
Working WIP exhaustive case macro

vendethiel commented on test/adt_test.exs:
```diff
@@ -54,6 +54,20 @@ defmodule AdtTest do
+    end
+    assert result == "foo"
+  end
+
```
please add a failing test with `catch_error(Code.compile_quoted do ... end)`

bobjflong replied:
done using `eval_string`

bobjflong committed:
add test for non-exhaustive case

vendethiel commented:
How about

```
AdtDefinition.case foo,
  Foo: fn(foo) -> end,
  Bar: fn(bar) -> end,
```

bobjflong commented:
I hope to take a look at that syntax soon
On 9 Jul 2016 5:49 p.m., "ven" notifications@github.com wrote:

> How about
> 
> AdtDefinition.case foo,
>   Foo: fn(foo) -> end,
>   Bar: fn(bar) -> end,
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/vendethiel/ADT.ex/pull/6#issuecomment-231543702, or mute
> the thread
> https://github.com/notifications/unsubscribe/ACKPQJa74hhAeOY0ryqNE2cFHfb_fZHFks5qT9EbgaJpZM4JGT1u
> .

bobjflong committed:
fixup and cleanup syntax

vendethiel commented on lib/adt.ex:
```diff
@@ -24,7 +24,38 @@ defmodule ADT do
+          unquote(values),
+          fn {variant, _} ->
+            # Maps a module like Foo.Bar.Baz into a short string "Baz"
+            Regex.named_captures(~r/\.(?<short>[^\.]+)$/, inspect(variant), include_captures: true) |> Map.fetch!("short")
```
maybe it'd be simpler to split-then-last? or at least make this a private helper

vendethiel commented on lib/adt.ex:
```diff
@@ -24,7 +24,38 @@ defmodule ADT do
+  defp case_macro(values) do
+    quote do
+      defmacro case(a, [do: statements]) do
+        possible_variants = Enum.map(
```
why not `unquote(values) |> Enum.map` like L40?

vendethiel commented on lib/adt.ex:
```diff
@@ -24,7 +24,38 @@ defmodule ADT do
+            Regex.named_captures(~r/\.(?<short>[^\.]+)$/, inspect(variant), include_captures: true) |> Map.fetch!("short")
+          end
+        ) |> Enum.sort
+        given_variants = statements |> Enum.map(fn {k, _} -> to_string(k) end) |> Enum.sort
```
seeing as we need `k` to be a string pretty much everywhere, would it make more sense to map it when we get it?

vendethiel commented on lib/adt.ex:
```diff
@@ -24,7 +24,38 @@ defmodule ADT do
+
+  defp case_macro(values) do
+    quote do
+      defmacro case(a, [do: statements]) do
```
why not... do away with `do`? (so we don't have the array wrapping)

bobjflong committed:
Better error message

bobjflong committed:
Better test for error case

bobjflong commented:
I think I've taken those comments on board. I've also cleaned up the code a bit and added a better test for the non-exhaustive case. I think a good next step would be to add a lot of test cases.

vendethiel commented:
Very nice work.

bobjflong committed:
remove one regex

vendethiel commented on lib/adt.ex:
```diff
@@ -56,12 +56,8 @@ defmodule ADT do
-  def _shorten_match(match) do
-    Regex.named_captures(~r/\.(?<short>[^\.]+){/, inspect(match), include_captures: true) |> Map.fetch!("short")
+  def _shorten(name) do
+    Regex.named_captures(~r/\.(?<short>[^\.{]+)($|{)/, inspect(name), include_captures: true) |> Map.fetch!("short")
```
The dot needs not be escaped in the character class(...or elixir's regexes are weird..)

vendethiel commented on lib/adt.ex:
```diff
@@ -24,7 +24,44 @@ defmodule ADT do
+          raise ADT._exhaustive_error(possible_variants, given_variant_names)
+        end
+        rules = Enum.flat_map(given_variants, fn {k, v} ->
+          condition = quote do
```
Now that this is only one line, why not do away with `condition` altogether and inline it right below?

bobjflong committed:
simplify based on feedback

vendethiel commented on lib/adt.ex:
```diff
@@ -57,7 +54,7 @@ defmodule ADT do
   # Maps a module like Foo.Bar.Baz into a short string "Baz"
   def _shorten(name) do
-    Regex.named_captures(~r/\.(?<short>[^\.{]+)($|{)/, inspect(name), include_captures: true) |> Map.fetch!("short")
+    Regex.named_captures(~r/.(?<short>[^.{]+)($|{)/, inspect(name), include_captures: true) |> Map.fetch!("short")
```
The first escape was okay (`~r/\.`). This still works, but backtracks a lot more than your original version.

bobjflong replied:
Do you think it's best to interpolate the terminator so that it becomes a simpler regex? Or maybe two regexes was ok

vendethiel replied:
I'm fine with that terminator (it could be `(?:$|{)` or whatever else). I was just talking about escaping the leading dot.

bobjflong replied:
ah sorry, I'll fix that.

vendethiel commented on test/adt_test.exs:
```diff
@@ -54,6 +54,29 @@ defmodule AdtTest do
+  end
+
+  test "non-exhaustive case" do
+    error = catch_error Code.eval_string """
```
any reason you use `eval_string` here instead of `eval_quoted` (that's used before)?

bobjflong replied:
It's weird but

``` elixir
    error = catch_error(Code.eval_quoted do
      require AdtTest.AdtDefinition

      result = AdtTest.AdtDefinition.case %AdtTest.AdtDefinition.Foo{}, [
        Foo: fn(x) -> "foo" end
      ]
    end)
```

Seems to throw an error in the test, and not catch. Something subtle to do with the order the arguments are evaluated maybe?

vendethiel replied:
Maybe we should go about it differently because of the compile-time raise? Not sure. I should ask on IRC...

vendethiel commented on test/adt_test.exs:
```diff
@@ -54,6 +54,29 @@ defmodule AdtTest do
+        Foo: fn(x) -> "foo" end
+      ]
+    """
+    assert error.message == "case macro not exhaustive.\nGiven [\"Foo\"].\nPossible: [\"Bar\", \"Foo\"]."
```
just an idea, not sure if it's better or worse... What about listing `you forgot to cover cases x, y, z`?

bobjflong replied:
It's definitely worth thinking about, one thing I'd like to cover is dead code. Given:

``` elixir
  defmodule AdtDefinition do
    ADT.define foo(a: 0) | bar(val: "hey")
  end
```

``` elixir
    result = AdtDefinition.case val, [
      Foo: fn(_) -> "1" end,
      Bar: fn(_) -> "2" end,
      Baz: fn(_) -> "3" end
    ]
```

I think this should _also_ raise, as the programmer has written dead code in the `Baz` case.

vendethiel replied:
You're right! I think this currently raises (though there's no test for it), it'd be a bad thing to lose that feature :-).

bobjflong replied:
👍 done

bobjflong committed:
changes based on feedback

vendethiel commented:
@bobjflong Do you want me to merge this now, or do you want to try to work on "Missing/extra cases: ..." in this iteration?

bobjflong commented:
Test cases? I'm happy to merge. I definitely will follow up with more test cases and improvements

vendethiel commented:
Okay. I can either merge it myself or give you the keys to merge it yourself, whichever you prefer?

bobjflong commented:
👍 I forgot you invited me to the project. Merging

vendethiel commented:
👍 thanks for your work!
